File: PracticasPDOOJava/Practica2_Java/src/irrgarten/Player.java
package irrgarten;

import java.util.ArrayList;

/**
 * Representa un jugador dentro del laberinto
 * <p>
 * Cada jugador posee características propias como inteligencia, fuerza y salud,
 * además de una posición (row/column) dentro del {@link labyrinth}.
 * Puede equipar un número de limitado de armas y escudos, atacar, defenderse y 
 * resucitar tras ser derrotado.
 * </p>
 * 
 * @author Joaquín Cruz Lorenzo
 */
public class Player {
    // ==================================
    // CONSTANTES DE CLASE
    // ==================================
    private static final int MAX_WEAPONS = 2;
    private static final int MAX_SHIELDS = 3;
    private static final int INITIAL_HEALTH = 10;
    private static final int HITS2LOSE = 3;
    
    // ==================================
    // ATRIBUTOS DE INSTANCIA 
    // ==================================
    private String name;
    private char number;
    private float intelligence, strength, health;
    private int row, col, consecutiveHits = 0;
    
    private ArrayList<Weapon> weapons = new ArrayList();
    private ArrayList<Shield> shields = new ArrayList();
    
    /**
     * Constructor
     * @param number
     * @param intelligence
     * @param strength 
     */
    public Player(char number, float intelligence, float strength) {
        this.number = number;
        this.intelligence = intelligence;
        this.strength = strength;
        this.health = INITIAL_HEALTH;
        this.row = -1;
        this.col = -1;
        this.name = "Player #" + number;
    }
    
    // ==================================
    // MÉTODOS CONSULTORES
    // ==================================

    /**
     * 
     * @return fila actual del jugador 
     */
    public int getRow() {
        return row;
    }
    
    /**
     * 
     * @return columna actual del jugador
     */
    public int getCol() {
        return col;
    }
    
    /**
     * 
     * @return número del jugador 
     */
    public char getNumber() {
        return number;
    }
    
    
     // ==================================
    // MÉTODOS DE CONTROL DE ESTADO
    // ==================================
    /**
     * Restaura al jugador al estado inicial.
     * <ul>
     *          <li> Vacía armas y escudos. </li>
     *          <li> Reinicia salud e impactos consecutivos. </li>
     * </ul>
     */
    public void resurrect() {
        weapons.clear();
        shields.clear();
        health = INITIAL_HEALTH;
        consecutiveHits = 0;
    }
    
    /**
     * Cambia la posicion del jugador dentro del laberinto.
     * @param row
     * @param col 
     */
    public void setPos(int row, int col) {
        this.row = row;
        this.col = col;
    }
    
    /**
     * Indica si el jugador está muerto (salud menor o igual a 0).
     * @return true si el jugador está muerto, false en caso contrario.
     */
    public boolean dead() {
       return health <= 0;
    }
    
    // ==================================
    // COMBATE
    // ==================================
    /**
     * Calucla la potencia de ataque del jugador.
     * <p> La fórmula básica es {@code strength + sumWeapons()} </p>
     * @return  potencia de ataque total
     */
    public float attack() {
        return strength + sumWeapons();
    }
    
    /**
     * Inicia la defensa ante un ataque recibido.
     * <p> Lo dejamos sin implementar en P2. Se nos dará información en P3 </p>
     * @param receivedAttack intensidad del ataque recibido
     * @return 
     */
    public boolean defend(float receivedAttack) {
        throw new UnsupportedOperationException();
    }
    
    /**
     * Suma la potencia de todas las armas equipadas (consumiendo sus usos)
     * @return suma total de ataque de armas activas
     */
    private float sumWeapons() {
        float total = 0.0f;
        
        for (Weapon w : weapons) total += w.attack();
        
        return total;
    }
    
    /**
     * Suma la capacidad de protección de todos los escudos (consumiendo sus usos)
     * @return  protección total ofrecida por los escudos
     */
    private float sumShields() {
        float total = 0.0f;
        
        for (Shield s : shields) total += s.protect();
        
        return total;
    }
    
    /**
     * Calcula la energía defensiva del jugador.
     * <p> {@code defensiveEnergy = intelligence + sumShields()} </p>
     * @return  valor de energía defensiva
     */
    private float defensiveEnergy() {
        return intelligence + sumShields();
    }
    
    /**
     * Incrementa el número de impactos consecutivos sufridos.
     */
    private void incConsecutiveHits() {
        consecutiveHits++;
    }
    
    /**
     * Reinicia el contador de impactos consecutivos a 0.
     */
    private void resetHits() {
        consecutiveHits = 0;
    }
    
    /**
     * Resta un punto de salud al jugador.
     */
    public void gotWounded() {
        health--;
    }
    
    // ==================================
    // EQUIPAMIENTO
    // ==================================
    /**
     * Genera un arma nueva con valores aleatorios dados por {@link Dice}.
     * @return  nueva instancia de {@link Weapon}
     */
    private Weapon newWeapon() {
        return new Weapon(Dice.weaponPower(), Dice.usesLeft());
    }
    
    /**
     * Genera un escudo nuevo con valores aleatorios dados por {@link Dice}.
     * @return nueva instancia de {@link Shield}
     */
    private Shield newShield() {
        return new Shield(Dice.shieldPower(), Dice.usesLeft());
    }
    
    /**
     * Añade un arma al inventario del jugador.
     * <p> En P2 no la implemento, la implemento en P3 </p>
     * @param w arma a añadir
     */
    private void receiveWeapon(Weapon w) {
        throw new UnsupportedOperationException();
    }
    
    /**
     * Añade un escudo al inventario del jugador.
     * <p> En P2 no la implemento, la implemento en P3 </p>
     * @param s escudo a añadir
     */
    private void receiveShield(Shield s) {
        throw new UnsupportedOperationException();
    }
    
    /**
     * Recibe una recompensa tras un combate ganado.
     * <p> En P2 no la implemento, la implemento en P3 </p>
     */
    public void receiveReward() {
        throw new UnsupportedOperationException();
    }
    
    /**
     * Gestiona el impacto recibido tras un combate ganado.
     * <p> En P2 no la implemento, la implemento en P3 </p>
     * @param receivedAttack ataque recibido
     * @return true si el jugador sobrevive, false en caso contrario
     */
    private boolean manageHit(float receivedAttack) {
        throw new UnsupportedOperationException();
    }
    
    /**
     * Movimiento del jugador
     * <p> En P2 no la implemento, la implemento en P3 </p>
     * @param direction         dirección preferida
     * @param validMoves     movimientos válidos posibles
     * @return  dirección final seleccionada
     */
    public Directions move(Directions direction, Directions[] validMoves){
        throw new UnsupportedOperationException();
    }
    
    // ==================================
    // REPRESENTACIÓN TEXTUAL
    // ==================================
    @Override
    public String toString() {
        return "Player{" + "name= " + name + ", number= " + number + ", intelligence=" + intelligence + ", strength= " + strength + ", health= " + health + 
                ", row= " + row + ", col= " + col + ", consecutiveHits= " + consecutiveHits + ", weapons= " + weapons + ", shields= " + shields + '}';
    }
}

File: PracticasPDOORuby/Practica4_Ruby/lib/Irrgarten/player.rb
# frozen_string_literal: true

require_relative 'labyrinth_character'
require_relative 'dice'
require_relative 'weapon'
require_relative 'shield'

module Irrgarten
  class Player < LabyrinthCharacter

    # Constantes
    MAX_WEAPONS = 2
    MAX_SHIELDS = 3
    INITIAL_HEALTH = 10
    HITS2LOSE = 3

    # Atributos específicos de Player (los demás están en el padre)
    attr_reader :number, :consecutive_hits, :weapons, :shields

    def initialize(number, intelligence, strength)
      # 1. Inicializamos la parte del padre
      super("Player #{number}", intelligence, strength, INITIAL_HEALTH)

      # 2. Inicializamos la parte propia
      @number = number
      @consecutive_hits = 0
      @weapons = []
      @shields = []
    end

    # NUEVO P4: Metodo copy (sustituye al constructor de copia)
    def copy(other)
      # Copiamos la parte del padre
      super(other)

      # Copiamos la parte propia
      @number = other.number
      @consecutive_hits = other.consecutive_hits

      # Copia profunda de armas y escudos
      @weapons = []
      @shields = []

      other.weapons.each do |w|
        @weapons << Weapon.new(w.effect, w.uses)
      end

      other.shields.each do |s|
        @shields << Shield.new(s.effect, s.uses)
      end
    end

    def resurrect
      self.health = INITIAL_HEALTH
      @consecutive_hits = 0
      @weapons = []
      @shields = []
    end

    # Movimiento: delega en el padre para saber la posición,
    # pero mantiene la lógica de dirección
    def move(direction, valid_moves)
      size = valid_moves.size
      contained = valid_moves.include?(direction)

      if size > 0 && !contained
        first = valid_moves[0]
        return first
      else
        return direction
      end
    end

    def attack
      @strength + sum_weapons
    end

    def defend(received_attack)
      manage_hit(received_attack)
    end

    # Recepción de premios (sin CardDeck en Ruby)
    def receive_reward
      w_reward = Dice.weapons_reward
      s_reward = Dice.shields_reward

      w_reward.times do
        wnew = new_weapon
        receive_weapon(wnew)
      end

      s_reward.times do
        snew = new_shield
        receive_shield(snew)
      end

      extra_health = Dice.health_reward
      self.health += extra_health
    end

    def to_s
      "Player #{@number} " + super + "\n" +
        "  Weapons: #{@weapons.join(', ')}\n" +
        "  Shields: #{@shields.join(', ')}"
    end

    # --- Métodos privados/protegidos ---
    private

    def receive_weapon(w)
      @weapons.each do |wi|
        discard = wi.discard
        if discard
          @weapons.delete(wi)
        end
      end

      size = @weapons.size
      if size < MAX_WEAPONS
        @weapons << w
      end
    end

    def receive_shield(s)
      @shields.each do |si|
        discard = si.discard
        if discard
          @shields.delete(si)
        end
      end

      size = @shields.size
      if size < MAX_SHIELDS
        @shields << s
      end
    end

    def new_weapon
      Weapon.new(Dice.weapon_power, Dice.uses_left)
    end

    def new_shield
      Shield.new(Dice.shield_power, Dice.uses_left)
    end

    def sum_weapons
      sum = 0
      @weapons.each { |w| sum += w.attack }
      sum
    end

    def sum_shields
      sum = 0
      @shields.each { |s| sum += s.protect }
      sum
    end

    def defensive_energy
      @intelligence + sum_shields
    end

    def manage_hit(received_attack)
      defense = defensive_energy
      if defense < received_attack
        got_wounded
        inc_consecutive_hits
      else
        reset_hits
      end

      lose = false
      if @consecutive_hits == HITS2LOSE || dead
        reset_hits
        lose = true
      end
      lose
    end

    def reset_hits
      @consecutive_hits = 0
    end

    def inc_consecutive_hits
      @consecutive_hits += 1
    end

    # Hacemos estos métodos protegidos para que FuzzyPlayer pueda usarlos
    protected :sum_weapons, :sum_shields, :defensive_energy
  end
end